### PR TITLE
Hf hub generative baselines and starter kit

### PR DESCRIPTION
**Pull request by uvafan**
Lots of restructuring to do this somewhat nicely.

No need to review closely, the most important parts are `src/raft_baselines/scripts/starter_kit.ipynb` and `transformers_causal_lm_classifier.py`.

TODOs:
1.  Test code for GPUs
2. Changes based on Lewis feedback on starter kit

If time:
1. Add an option for using the inference API
2. Add zero-shot classificaiton pipeline using library and/or inference API
3. Clean up types some, I think a few are probably wrong

### CODE REVIEW

File: src/raft_baselines/classifiers/transformers_causal_lm_classifier.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import torch
+from transformers import AutoModelForCausalLM
+from sentence_transformers import util
+
+from raft_baselines.classifiers.in_context_classifier import InContextClassifier
+from raft_baselines.utils.tokenizers import TransformersTokenizer
+from raft_baselines.utils.embedders import SentenceTransformersEmbedder
+
+
+class TransformersCausalLMClassifier(InContextClassifier):
```
Comment (lewtun) [inline src/raft_baselines/classifiers/transformers_causal_lm_classifier.py:14]: This looks really clean - thanks for implementing it! 

One thing that would help my understanding is a small code snippet (or maybe a script?) that explains the steps one has to take to go from the RAFT dataset to a set of predictions, e.g. I'm thinking of something like

```python
from datasets import load_dataset
from raft_baselines import TransformersCausalLMClassifier

raft_dset = load_dataset("ought/raft", "banking_77")
clf = TransformersCausalLMClassifier()
# what do I do next?
```

File: src/raft_baselines/classifiers/transformers_causal_lm_classifier.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        self.similarity_embedder = SentenceTransformersEmbedder()
+
+        super().__init__(
+            *args,
+            tokenizer=tokenizer,
+            max_tokens=self.model.config.max_position_embeddings,
+            **kwargs,
+        )
+
+    def semantically_select_training_examples(
```
Comment (lewtun) [inline src/raft_baselines/classifiers/transformers_causal_lm_classifier.py:33]: Just an idea, but if you didn't want to introduce the `sentence-transformers` dependency, you can do semantic search directly in `datasets`: https://huggingface.co/docs/datasets/faiss_and_ea.html#adding-a-faiss-index

Not strong feeling on this, so feel free to ignore :)

**Review by lewtun — COMMENTED** [review]
Thanks for adding the baseline - the implementation looks nice 🚀 !

I left two small suggestions, but the general structure looks good IMO

File: src/raft_baselines/classifiers/transformers_causal_lm_classifier.py (hunk 1 above)
Comment (uvafan) [reply src/raft_baselines/classifiers/transformers_causal_lm_classifier.py:14]: [raft_predict.py](https://github.com/oughtinc/raft-baselines/blob/hf-hub-generative-baselines/src/raft_baselines/scripts/raft_predict.py) is the script which runs classifiers on predictions; the forthcoming starter kit notebook should contain the code for this as well

File: src/raft_baselines/classifiers/transformers_causal_lm_classifier.py (hunk 2 above)
Comment (uvafan) [reply src/raft_baselines/classifiers/transformers_causal_lm_classifier.py:33]: I'll leave as is for now

Comment (lewtun) [reply src/raft_baselines/classifiers/transformers_causal_lm_classifier.py:33]: sounds good!

File: src/raft_baselines/classifiers/in_context_classifier.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        prompt: str,
+    ) -> List[float]:
+        ...
+
+    def _classify_prompt(
+        self,
+        prompt: str,
+    ) -> Dict[str, float]:
+        raw_p = self._get_raw_probabilities(prompt)
+        raw_p = torch.stack(raw_p).cpu().detach().numpy()
```
Comment (lewtun) [inline src/raft_baselines/classifiers/in_context_classifier.py:228]: I had to convert the list of tensors to a numpy array for cases when we run the code on a GPU

File: src/raft_baselines/classifiers/transformers_causal_lm_classifier.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+        # adapted from https://towardsdatascience.com/semantic-similarity-using-transformers-8f3cb5bf66d6
+        target_embedding = self.similarity_embedder(tuple([formatted_target]))
+        example_embeddings = self.similarity_embedder(formatted_examples_without_labels)
+
+        similarity_scores = util.pytorch_cos_sim(target_embedding, example_embeddings)[
+            0
+        ]
+
+        sorted_indices = torch.argsort(-similarity_scores.to(self.device))
```
Comment (lewtun) [inline src/raft_baselines/classifiers/transformers_causal_lm_classifier.py:52]: Here I changed `np.argsort` to `torch.argsort` so we can support sorting tensors on both CPU / GPU

**Comment by uvafan** [comment]
Thanks for the help @lewtun! I'm gonna go ahead and merge

**Pull request merged**